File: ProblemSolving/450DSA/Java/src/main/java/org/redquark/techinterview/dsa/binarytree/RightViewOfATree.java
package org.redquark.techinterview.dsa.binarytree;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

/**
 * @author Anirudh Sharma
 * <p>
 * Given a Binary Tree, find Right view of it.
 * Right view of a Binary Tree is set of nodes visible when tree is viewed from right side.
 */
public class RightViewOfATree {

    private static List<Integer> rightView(Node root) {
        // List to the result
        List<Integer> result = new ArrayList<>();
        // Special case
        if (root == null) {
            return result;
        }
        // Queue to store the nodes of the tree
        Queue<Node> nodes = new LinkedList<>();
        // Add root node to the queue
        nodes.add(root);
        // Loop until the queue is empty
        while (!nodes.isEmpty()) {
            // Get the size of the queue
            int size = nodes.size();
            // Loop until the size
            for (int i = 0; i < size; i++) {
                // Get the current node
                Node current = nodes.poll();
                // If this is the last node of
                // the current level, add it to
                // the result
                if (current != null && i == size - 1) {
                    result.add(current.data);
                }
                // Move to the left child
                if (current != null && current.left != null) {
                    nodes.add(current.left);
                }
                // Move to the right child
                if (current != null && current.right != null) {
                    nodes.add(current.right);
                }
            }
        }
        return result;
    }

    public static void main(String[] args) {
        Node root = new Node(1);
        root.left = new Node(3);
        root.right = new Node(2);
        System.out.println(rightView(root));

        root = new Node(10);
        root.left = new Node(20);
        root.right = new Node(30);
        root.left.left = new Node(40);
        root.left.right = new Node(60);
        System.out.println(rightView(root));


        root = new Node(1);
        root.left = new Node(2);
        root.right = new Node(3);
        root.left.left = new Node(4);
        root.left.right = new Node(5);
        System.out.println(rightView(root));
    }

    static class Node {
        final int data;
        Node left;
        Node right;

        Node(int data) {
            this.data = data;
        }
    }
}

File: ProblemSolving/450DSA/Python/src/binarytree/RightViewOfATree.py
"""
@author Anirudh Sharma

Given a Binary Tree, print right view of it.
Left view of a Binary Tree is set of nodes visible when tree is visited from right side.
"""


def rightView(root):
    # list to store the result
    result = []
    # Special case
    if root is None:
        return result
    # Queue to store the nodes of the tree
    nodes = []
    # Add the root node in the queue
    nodes.append(root)
    # Loop until the queue is empty
    while nodes:
        # Get the size of the queue
        size = len(nodes)
        # Loop until the size
        for i in range(size):
            # Get the current node
            current = nodes.pop(0)
            # If it is the first node at the current level,
            # add it to the result
            if i == size - 1:
                result.append(current.data)
            # Move to left child if exists
            if current.left:
                nodes.append(current.left)
            # Move to right child if exists
            if current.right:
                nodes.append(current.right)
    return result


class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


if __name__ == "__main__":
    root = Node(1)
    root.left = Node(3)
    root.right = Node(2)
    print(rightView(root))

    root = Node(10)
    root.left = Node(20)
    root.right = Node(30)
    root.left.left = Node(40)
    root.left.right = Node(60)
    print(rightView(root))

    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.left = Node(4)
    root.left.right = Node(5)
    print(rightView(root))
